File: single-language-analysis/dataset-files/java/TestSmell093.java
import org.junit.Test;
import static org.junit.Assert.*;

public class TestSmell093 {

    @Test
    public void testMinibatchUpdateConsistency() {
        // Check that dense and sparse minibatch update give the same results
        Random rng = new Random(globalRandomSeed);

        double[][] centersOld = centers.clone();
        for (int i = 0; i < centersOld.length; i++) {
            centersOld[i] = centersOld[i] + rng.nextGaussian();
        }
        double[][] centersOldCsr = centersOld.clone();

        double[][] centersNew = new double[centersOld.length][];
        double[][] centersNewCsr = new double[centersOldCsr.length][];

        double[] weightSums = new double[centersOld.length];
        double[] weightSumsCsr = new double[centersOldCsr.length];

        double[] sampleWeight = new double[X.length];
        Arrays.fill(sampleWeight, 1.0);

        // extract a small minibatch
        double[][] XMb = Arrays.copyOfRange(X, 0, 10);
        double[][] XMbCsr = Arrays.copyOfRange(XCsr, 0, 10);
        double[] sampleWeightMb = Arrays.copyOfRange(sampleWeight, 0, 10);

        // step 1: compute the dense minibatch update
        double oldInertia = miniBatchStep(
            XMb,
            sampleWeightMb,
            centersOld,
            centersNew,
            weightSums,
            new Random(globalRandomSeed),
            false
        );
        assertTrue(oldInertia > 0.0);

        // compute the new inertia on the same batch to check that it decreased
        double[] labels = new double[0];
        double newInertia = labelsInertia(XMb, sampleWeightMb, centersNew);
        assertTrue(newInertia > 0.0);
        assertTrue(newInertia < oldInertia);

        // step 2: compute the sparse minibatch update
        double oldInertiaCsr = miniBatchStep(
            XMbCsr,
            sampleWeightMb,
            centersOldCsr,
            centersNewCsr,
            weightSumsCsr,
            new Random(globalRandomSeed),
            false
        );
        assertTrue(oldInertiaCsr > 0.0);

        // compute the new inertia on the same batch to check that it decreased
        double[] labelsCsr = new double[0];
        double newInertiaCsr = labelsInertia(XMbCsr, sampleWeightMb, centersNewCsr);
        assertTrue(newInertiaCsr > 0.0);
        assertTrue(newInertiaCsr < oldInertiaCsr);

        // step 3: check that sparse and dense updates lead to the same results
        assertArrayEquals(labels, labelsCsr);
        assertArrayEquals(centersNew, centersNewCsr);
        assertEquals(oldInertia, oldInertiaCsr, 0.0);
        assertEquals(newInertia, newInertiaCsr, 0.0);
    }
}

File: single-language-analysis/dataset-files/python/test_smell_093.py
# Original URL: https://github.com/ogrisel/scikit-learn/blob/b20998d4ec606558ae86cba7ff5a84aea094c397/sklearn/cluster/tests/test_k_means.py#L205-L266

@pytest.mark.parametrize("X_csr", X_as_any_csr)
def test_smell_093(X_csr, global_random_seed):
    # Check that dense and sparse minibatch update give the same results
    rng = np.random.RandomState(global_random_seed)

    centers_old = centers + rng.normal(size=centers.shape)
    centers_old_csr = centers_old.copy()

    centers_new = np.zeros_like(centers_old)
    centers_new_csr = np.zeros_like(centers_old_csr)

    weight_sums = np.zeros(centers_old.shape[0], dtype=X.dtype)
    weight_sums_csr = np.zeros(centers_old.shape[0], dtype=X.dtype)

    sample_weight = np.ones(X.shape[0], dtype=X.dtype)

    # extract a small minibatch
    X_mb = X[:10]
    X_mb_csr = X_csr[:10]
    sample_weight_mb = sample_weight[:10]

    # step 1: compute the dense minibatch update
    old_inertia = _mini_batch_step(
        X_mb,
        sample_weight_mb,
        centers_old,
        centers_new,
        weight_sums,
        np.random.RandomState(global_random_seed),
        random_reassign=False,
    )
    assert old_inertia > 0.0

    # compute the new inertia on the same batch to check that it decreased
    labels, new_inertia = _labels_inertia(X_mb, sample_weight_mb, centers_new)
    assert new_inertia > 0.0
    assert new_inertia < old_inertia

    # step 2: compute the sparse minibatch update
    old_inertia_csr = _mini_batch_step(
        X_mb_csr,
        sample_weight_mb,
        centers_old_csr,
        centers_new_csr,
        weight_sums_csr,
        np.random.RandomState(global_random_seed),
        random_reassign=False,
    )
    assert old_inertia_csr > 0.0

    # compute the new inertia on the same batch to check that it decreased
    labels_csr, new_inertia_csr = _labels_inertia(
        X_mb_csr, sample_weight_mb, centers_new_csr
    )
    assert new_inertia_csr > 0.0
    assert new_inertia_csr < old_inertia_csr

    # step 3: check that sparse and dense updates lead to the same results
    assert_array_equal(labels, labels_csr)
    assert_allclose(centers_new, centers_new_csr)
    assert_allclose(old_inertia, old_inertia_csr)
    assert_allclose(new_inertia, new_inertia_csr)